polynote-server/src/main/scala/polynote/server/repository/fs/S3Filesystem.scala

```diff
@@ -0,0 +1,192 @@
+    */
+  override def writeStringToPath(path: Path, content: String): RIO[BaseEnv, Unit] = for {
+    bucket <- getBucket(path)
+    key: String = getKeyFromPath(path)
+    objToPut = new S3Object(bucket, key, content)
+    _ <- effectBlocking(service.putObject(bucket, objToPut)).uninterruptible
+  } yield ()
+
+
+  // TODO still need to reimplement
+  override def createLog(path: Path): RIO[BaseEnv, WAL.WALWriter] = {
+    effectBlocking(path.getParent.toFile.mkdirs()) *> FileChannelWALWriter(path)
+  }
+
+
+  override def list(path: Path): RIO[BaseEnv, List[Path]] = {
+    for {
+      bucket <- getBucket(path)
+      ret <- effectBlocking(service.listObjects(bucket, getKeyFromPath(path), "/"))
+    } yield ret.toList.map(objectToPath)
+  }
+
+  override def validate(path: Path): RIO[BaseEnv, Unit] = {
+    ZIO.unit
+    //    if (path.iterator().asScala.length > maxDepth) {
+    //      ZIO.fail(new IllegalArgumentException(s"Input path ($path) too deep, maxDepth is $maxDepth"))
+    //    } else ZIO.unit
+  }
+
+  override def exists(path: Path): RIO[BaseEnv, Boolean] = {
+    for {
+      bucket <- getBucket(path)
+      _ <- effectBlocking(service.getObjectDetails(bucket, getKeyFromPath(path)))
+    } yield true
+  }.catchSome {
+    case e: S3ServiceException =>
+      if (e.getResponseCode == 404) {
+        ZIO.succeed(false)
+      } else {
+        ZIO.fail(e)
+      }
+  }
+
+  override def move(from: Path, to: Path): RIO[BaseEnv, Unit] = {
+    for {
+      fromBucket <- getBucket(from)
+      toBucket <- getBucket(to)
+      _ <- effectBlocking(service.moveObject(
+        fromBucket.getName,
+        getKeyFromPath(from),
+        toBucket.getName,
+        new S3Object(getKeyFromPath(to)),
+        true))
+    } yield ()
+  }
+
+  override def copy(from: Path, to: Path): RIO[BaseEnv, Unit] = {
+    for {
+      fromBucket <- getBucket(from)
+      toBucket <- getBucket(to)
+      _ <- effectBlocking(
+        service.copyObject(
+          fromBucket.getName,
+          getKeyFromPath(from),
+          toBucket.getName,
+          new S3Object(getKeyFromPath(to)),
+          true
+        )
+      )
+    } yield ()
+  }
+
+  override def delete(path: Path): RIO[BaseEnv, Unit] = {
+    for {
+      bucket <- getBucket(path)
+      _ <- effectBlocking(service.getObjectDetails(bucket, getKeyFromPath(path)))
+      _ <- effectBlocking(service.deleteObject(bucket, getKeyFromPath(path)))
+    } yield ()
+  }
+
+  private def readBytes(is: => InputStream): RIO[BaseEnv, Chunk.Bytes] = {
+    for {
+      env <- ZIO.environment[BaseEnv]
+      ec = env.get[Blocking.Service].blockingExecutor.asEC
+      chunks <- fs2.io.readInputStream[Task](effectBlocking(is).provide(env), 8192, ec, closeAfterUse = true).compile.toChunk.map(_.toBytes)
+    } yield chunks
+  }
+
+  override def init(path: Path): RIO[BaseEnv, Unit] = ZIO.unit
+
+  def getBucket(p: Path): RIO[BaseEnv, S3Bucket] = {
```

Comment:
Should the bucket (and optionally a base path) be part of the notebook repository instead? Then, paths would just be paths. You'd parameterize the filesystem with the bucket, and create a new type of `NotebookRepository` which is configured with the base S3 URI.